core/Types.h

```diff
@@ -277,12 +277,18 @@ template <class To> To *cast_type(TypePtr &what) {
     return const_cast<To *>(cast_type<To>(static_cast<const TypePtr &>(what)));
 }
 
+template <class To> To *cast_type(TypePtr &&what) = delete;
+
 template <class To>
 typename TypePtr::TypeToCastType<To, TypePtr::TypeToIsInlined<To>::value>::type cast_type_nonnull(const TypePtr &what) {
     ENFORCE_NO_TIMER(isa_type<To>(what));
     return *reinterpret_cast<const To *>(what.get());
 }
 
+template <class To>
+typename TypePtr::TypeToCastType<To, TypePtr::TypeToIsInlined<To>::value>::type
+cast_type_nonnull(TypePtr &&what) = delete;
+
```

Comment:
Can we add a comment here saying this is deleted because it's too easy to accidentally have the TypePtr destruct before you have a chance to do something interesting with the casted result?